packages/hydrogen/src/cache/create-with-cache.ts

```diff
@@ -18,45 +20,82 @@ type CreateWithCacheOptions = {
+) {
   const {cache, waitUntil, request} = cacheOptions;
 
-  return function withCache<T = unknown>(
-    cacheKey: CacheKey,
-    strategy: CachingStrategy,
-    actionFn: ({addDebugData}: CacheActionFunctionParam) => T | Promise<T>,
-  ) {
-    return runWithCache<T>(cacheKey, actionFn, {
-      strategy,
-      cacheInstance: cache,
-      waitUntil,
-      debugInfo: {
-        ...getDebugHeaders(request),
-        stackInfo: getCallerStackLine?.(),
-      },
-    });
+  return {
+    /**
+     * Utility function that executes asynchronous operations and caches the
+     * result according to the strategy provided. Use this to do any type
+     * of asynchronous operation where `withCache.fetch` is insufficient.
+     * For example, when making multiple calls to a third-party API where the
+     * result of all of them needs to be cached under the same cache key.
+     * Whatever data is returned from the `actionFn` will be cached according
+     * to the strategy provided.
+     * Use the `CachingStrategy` to define a custom caching mechanism for your data.
+     * Or use one of the built-in caching strategies: `CacheNone`, `CacheShort`, `CacheLong`.
+     * > Note:
+     * > To prevent caching the result you must throw an error. Otherwise, the result will be cached.
+     * > For example, if you call `fetch` but the response is not successful (e.g. status code >= 400),
+     * > you should throw an error. Otherwise, the response will be cached.
+     */
+    run<InferredActionReturn = T>(
+      cacheKey: CacheKey,
+      strategy: CachingStrategy,
+      actionFn: ({
+        addDebugData,
+      }: CacheActionFunctionParam) =>
+        | InferredActionReturn
+        | Promise<InferredActionReturn>,
+    ) {
+      return runWithCache(cacheKey, actionFn, {
+        strategy,
+        cacheInstance: cache,
+        waitUntil,
+        debugInfo: {
+          ...getDebugHeaders(request),
+          stackInfo: getCallerStackLine?.(),
+        },
+      });
+    },
+
+    /**
+     * Fetches data from a URL and caches the result according to the strategy provided.
+     * When the response is not successful (e.g. status code >= 400), the caching is
+     * skipped automatically and the returned `data` is `null`.
+     * You can also prevent caching by using the `shouldCacheResponse` option and returning
+     * `false` from the function you pass in. For example, you might want to fetch data from a
+     * third-party GraphQL API but not cache the result if the GraphQL response body contains errors.
+     */
+    fetch<Body = T>(
+      url: string,
+      requestInit?: RequestInit,
+      options?: Pick<DebugOptions, 'displayName'> &
+        Pick<
+          FetchCacheOptions<Body>,
+          'cache' | 'cacheKey' | 'shouldCacheResponse'
+        >,
+    ): Promise<{data: Body | null; response: Response}> {
+      return fetchWithServerCache<Body | null>(url, requestInit ?? {}, {
+        waitUntil,
+        cacheKey: [url, requestInit],
```

Comment:
I think this `requestInit` might end up as `[Object object]`

Reply:
mmm ... if we are debating over what is safer between never hit a cache read or cache bleed, I would say never hit a cache is safer.

Then I'm fine with spreading the requestInit and the header as part of the cache key